src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityWorkbench.java

```diff
@@ -198,34 +196,246 @@ public void clearMachineInventory(@NotNull List<@NotNull ItemStack> itemBuffer)
+                        .marginTop(18)
+                        .background(GTGuiTextures.SLOT.asIcon().size(22))
+                        .marginBottom(4))
+                .child(IKey.dynamic(() -> TextFormattingUtil.formatLongToCompactString(amountCrafted.getIntValue(), 5))
+                        .alignment(Alignment.Center)
+                        .asWidget().widthRel(1f));
+    }
+
+    public IWidget createRecipeMemoryGrid(PanelSyncManager syncManager) {
+        return SlotGroupWidget.builder()
+                .matrix("XXX",
+                        "XXX",
+                        "XXX")
+                .key('X', i -> new RecipeMemorySlot(this.recipeMemory, i)
+                        .background(GTGuiTextures.SLOT))
+                .build().right(0);
+    }
+
+    public IWidget createInventoryPage(PanelSyncManager syncManager) {
+        if (this.connectedInventory.getSlots() == 0) {
+            return Flow.column()
+                    .debugName("inventory page - empty")
+                    .leftRel(0.5f)
+                    .padding(2)
+                    .height(18 * 6)
+                    .width(18 * 8 + 4)
+                    .background(GTGuiTextures.DISPLAY);
+        }
+
+        // this is actually supposed to include the tool and storage inventory
+        // but that causes problems
+        List<ItemSlot> list = new ArrayList<>(this.connectedInventory.getSlots());
+
+        int rowSize = Math.min(this.connectedInventory.getSlots(), 8);
+        var connected = new SlotGroup("connected_inventory", rowSize, true)
+                .setAllowSorting(false);
+        syncManager.registerSlotGroup(connected);
+
+        for (int i = 0; i < this.connectedInventory.getSlots(); i++) {
+            list.add(new ItemSlot()
+                    .setEnabledIf(itemSlot -> {
+                        int slot = itemSlot.getSlot().getSlotIndex();
+                        return slot < this.connectedInventory.getSlots();
+                    })
+                    .slot(trackSlot(this.connectedInventory, i)
+                            .slotGroup(connected)));
+        }
+
+        // sort list
+        list.sort((o1, o2) -> {
+            var left = o1.getSlot().getStack();
+            var right = o2.getSlot().getStack();
+
+            if (!left.isEmpty() && !right.isEmpty()) return 0;
+            if (left.isEmpty() && right.isEmpty()) return 0;
+
+            return right.isEmpty() ? -1 : 1;
+        });
+
+        return Flow.column()
+                .debugName("inventory page")
+                .padding(2)
+                .leftRel(0.5f)
+                .coverChildren()
+                .background(GTGuiTextures.DISPLAY)
+                .child(new Grid()
+                        .scrollable(new VerticalScrollData())
+                        .width(18 * 8 + 4)
+                        .height(18 * 6)
+                        .mapTo(rowSize, list));
+    }
+
+    public void sendHandlerToClient(PacketBuffer buffer) {
```

Comment:
This just sends the connected inventories, but the readHandler method uses all the inventories. Do we need to write the internal inventory and the tool inventory as well?